nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,599 @@
+{
+    out << path << " = ";
+    printValue(ctx, out, std::move(v), path);
+    out << ";\n";
+}
+
+void printAll(Context * ctx, Out & out)
+{
+    mapOptions(
+        [ctx, &out](std::string const & option_path) {
+            mapConfigValuesInOption(
+                [ctx, &out](std::string const & config_path, std::variant<Value, std::exception_ptr> v) {
+                    printConfigValue(ctx, out, config_path, v);
+                },
+                option_path, ctx);
+        },
+        ctx, ctx->options_root);
+}
+
+void printAttr(Context * ctx, Out & out, std::string const & path, Value * root)
+{
+    try {
+        printValue(ctx, out, *findAlongAttrPath(*ctx->state, path, *ctx->autoArgs, *root), path);
+    } catch (Error & e) {
+        out << describeError(e);
+    }
+}
+
+void printOption(Context * ctx, Out & out, std::string const & path, Value * option)
+{
+    out << "Value:\n";
+    printAttr(ctx, out, path, &ctx->config_root);
+
+    out << "\n\nDefault:\n";
+    printAttr(ctx, out, "default", option);
+
+    out << "\n\nType:\n";
+    printAttr(ctx, out, "type.description", option);
+
+    out << "\n\nExample:\n";
+    printAttr(ctx, out, "example", option);
+
+    out << "\n\nDescription:\n";
+    printAttr(ctx, out, "description", option);
+
+    out << "\n\nDeclared by:\n";
+    printAttr(ctx, out, "declarations", option);
+
+    out << "\n\nDefined by:\n";
+    printAttr(ctx, out, "files", option);
+    out << "\n";
+}
+
+void printListing(Out & out, Value * v)
+{
+    // Print this header on stderr rather than stdout because the old shell script
+    // implementation did.  I don't know why.
+    std::cerr << "This attribute set contains:\n";
+    for (const auto & a : v->attrs->lexicographicOrder()) {
+        std::string name = a->name;
+        if (!name.empty() && name[0] != '_') {
+            out << name << "\n";
+        }
+    }
+}
+
+bool optionTypeIs(Context * ctx, Value & v, std::string const & sought_type)
+{
+    try {
+        auto const & type_lookup = v.attrs->find(ctx->state->sType);
+        if (type_lookup == v.attrs->end()) {
+            return false;
+        }
+        Value type = evaluateValue(ctx, type_lookup->value);
+        if (type.type != tAttrs) {
+            return false;
+        }
+        auto const & name_lookup = type.attrs->find(ctx->state->sName);
+        if (name_lookup == type.attrs->end()) {
+            return false;
+        }
+        Value name = evaluateValue(ctx, name_lookup->value);
+        if (name.type != tString) {
+            return false;
+        }
+        return name.string.s == sought_type;
+    } catch (Error &) {
+        return false;
+    }
+}
+
+Value getSubOptions(Context * ctx, Value & option)
+{
+    Value getSubOptions =
+        evaluateValue(ctx, findAlongAttrPath(*ctx->state, "type.getSubOptions", *ctx->autoArgs, option));
+    if (getSubOptions.type != tLambda) {
+        throw Error("Option's type.getSubOptions isn't a function");
```

Comment:
I'm guessing this is a pretty rare/obscure error, but when it's encountered it would be helpful to know which option (attr path) this applies to.

Reply:
Done